eval/src/vespa/eval/tensor/dense/dense_tensor_apply.hpp

```diff
@@ -11,18 +11,17 @@ namespace vespalib::tensor::dense {
 template <typename Function>
 std::unique_ptr<Tensor>
 apply(DenseTensorAddressCombiner & combiner, DirectDenseTensorBuilder & builder,
-      const DenseTensorView &lhs, AddressContext & rhsAddr, const DenseTensorView::CellsRef & rhsCells,
-      Function &&func) __attribute__((noinline));
+      const DenseTensorView &lhs, const DenseTensorView::CellsRef & rhsCells, Function &&func) __attribute__((noinline));
 
 template <typename Function>
 std::unique_ptr<Tensor>
 apply(DenseTensorAddressCombiner & combiner, DirectDenseTensorBuilder & builder,
-      const DenseTensorView &lhs, AddressContext & rhsAddr, const DenseTensorView::CellsRef & rhsCells, Function &&func)
+      const DenseTensorView &lhs, const DenseTensorView::CellsRef & rhsCells, Function &&func)
 {
     for (DenseTensorCellsIterator lhsItr = lhs.cellsIterator(); lhsItr.valid(); lhsItr.next()) {
         combiner.updateLeftAndCommon(lhsItr.address());
-        if (rhsAddr.updateCommon(combiner.address(), combiner.commonRight())) {
-            combiner.for_each(rhsAddr, rhsCells, [&func, &builder, &lhsItr](size_t combined, double rhsCell) {
+        if (combiner.updateCommon()) {
+            combiner.for_each_right(rhsCells, [&func, &builder, &lhsItr](size_t combined, double rhsCell) {
```

Comment:
I got a new idea here:
since we let the lambda handle the left iterator, we could also let it handle the right cell lookup;
passing the rhsCells to the lambda itself and letting the for_each_right function pass indexes for both combined and right cells will separate address combining into the combiner and keep cell referencing and lookup to the lambda function.